homeassistant/components/sensor/folder.py

```diff
@@ -58,21 +85,54 @@ class Folder(Entity):
 
     ICON = 'mdi:folder'
 
-    def __init__(self, folder_path, filter_term):
+    def __init__(self, folder_path, filter_term, name, recursive):
         """Initialize the data object."""
         folder_path = os.path.join(folder_path, '')  # If no trailing / add it
         self._folder_path = folder_path   # Need to check its a valid path
         self._filter_term = filter_term
-        self._number_of_files = None
-        self._size = None
-        self._name = os.path.split(os.path.split(folder_path)[0])[1]
+        if name == DEFAULT_NAME:
+            self._name = os.path.split(os.path.split(folder_path)[0])[1]
+        else:
+            self._name = name
+        self._recursive = recursive
+        self._files_record = get_files_dict(
+            folder_path, filter_term, recursive)
+        self._number_of_files = len(self._files_record)
+        self._size = get_size(list(self._files_record.keys()))
         self._unit_of_measurement = 'MB'
+        self._last_added = None
+        self._last_deleted = None
+        self._last_modified = None
 
     def update(self):
         """Update the sensor."""
-        files_list = get_files_list(self._folder_path, self._filter_term)
-        self._number_of_files = len(files_list)
-        self._size = get_size(files_list)
+        current_files = get_files_dict(
+            self._folder_path, self._filter_term, self._recursive)
+        self._number_of_files = len(current_files)
+        self._size = get_size(list(current_files.keys()))
+
+        for file_path in set(
```

Comment:
This is SUPER Inefficient. Operating systems have events that one can listen to. You can do that using the PyPi package `watchdog`

Reply:
The functionality in its current implementation cannot be merged.